helm: allow to disable local auth in teleport-cluster chart

hugoShaka opened the pull request:
Fixes #12974 by adding a `disableLocalAuthentication` value.

hugoShaka committed:
helm: allow to disable local auth + nest auth values (#12974)

This commit introduces a new `authentication` value controlling:
- the authentication type (previously `authenticationType`)
- the second-factor (previously under `authenticationSecondFactor`)
- if local auth is enabled (new feature)
- the connector name (new feature)
- the locking mode (new feature)

If previous values `authenticationType` and `authenticationSecondFactor`
are specified we'll prefer them for backward compatibility.

marcoandredinis commented on docs/pages/kubernetes-access/helm/reference/teleport-cluster.mdx:
```diff
@@ -113,6 +113,17 @@ This field allows you to forbid specific device models and vendors, while allowi
 Devices within this list will be rejected during registration.
 By default no devices are forbidden.
 
+## `disableLocalAuthentication`
```
Is the name a requirement?
I usually try to avoid a double negative
In this case we have `disable` and a boolean, so we might have (it's actually the default):
`disableLocalAuthentication: false`

I think we should instead have a `localAuthentication` as setting name, and a default to `true`

webvictim replied:
I agree that a double negative is a bad idea.

zmb3 replied:
+1

r0mant replied:
I talked about this with Hugo and suggested to use "disable" notation here because of the potential issue with how Helm treats boolean fields. It looked like we can't easily distinguish between "false" and empty value which means that "localAuth: false" and absence of "localAuth" would result in a different behavior which is a little confusing. So in the config template we won't really know whether to set local auth to "false" or not include it (to get default behavior).

Hence, the suggestion to use the "disableLocalAuth". I agree double negation is not ideal though. Maybe it's not such a big deal since we control our Helm chart and will just have this value always set in it with explicit value.

webvictim replied:
Yep - we can set `enableLocalAuth: true` in the default `values.yaml` which will be included with all installs and someone can override it to `enableLocalAuth: false` to get this behaviour. I don't really like this either (surprise 😅) but it's definitely a lesser evil.

marcoandredinis replied:
We control the default value, so I think we can use `localAuth` and set the appropriate default

hugoShaka replied:
I removed the double negation and the chart now always sets the setting, the executable's default have no impact.

https://github.com/gravitational/teleport/pull/13827/commits/7051356984a68c95782d28672073dc5f61d99ca1#diff-fa006d6bafda53ef7d3c1ac89b897e70648bf64a0f3f5fc5489250e0c1dc5e7dR21-R29

Does this looks OK to you?

webvictim commented on examples/chart/teleport-cluster/.lint/disable-local-authentication.yaml:
```diff
@@ -0,0 +1,3 @@
+clusterName: helm-lint
+authenticationType: github
+disableLocalAuthentication: true
```
I'm inclined to suggest that we change this to a nested YAML structure like:
```yaml
authentication:
  type: github
  local_auth: false
```

That way it nests all authentication-related settings under one header and makes things easier to follow, and will be easier to extend in future if needed.

It'd need some logic to coalesce `authenticationType` when set for backwards compatibility.

hugoShaka replied:
Done. I left `authenticationSecondFactor` out of the `authentication` object as this implies a bigger refacto. I'm not sure if current state is OK, if I should continue the refacto here or move to a new issue. WDYT?

webvictim replied:
As we're introducing a new `authentication` top level object here, I'd lean towards also refactoring the `authenticationSecondFactor` item out (and adding a `coalesce` for backwards compatibility) at the same time. It's ultimately up to @r0mant though!

ptgott commented:
Hi @hugoShaka, would you mind rebasing this so it includes the commit that moves the Helm references? I'll give this a look after that.

hugoShaka commented:
> Hi @hugoShaka, would you mind rebasing this so it includes the commit that moves the Helm references? I'll give this a look after that.

Done

webvictim commented on docs/pages/setup/helm-reference/teleport-cluster.mdx:
```diff
@@ -63,13 +63,26 @@ This reference details available values for the `teleport-cluster` chart.
 | `string` | `local` | Yes | `auth_service.authentication.type` | ❌ |
 
-`authenticationType` controls the authentication scheme used by Teleport. Possible values are `local` and `github` for OSS, plus `oidc`, `saml`, and `false` for Enterprise.
+`authentication.type` controls the authentication scheme used by Teleport. Possible values are `local` and `github` for OSS, plus `oidc`, `saml` for Enterprise.
```
```suggestion
`authentication.type` controls the authentication scheme used by Teleport. Possible values are `local` and `github` for OSS, plus `oidc` and `saml` for Enterprise.
```

webvictim replied:
```suggestion
`authentication.type` controls the authentication scheme used by Teleport. Possible values are `local` and `github` for OSS, plus `oidc` and `saml` for Enterprise.
```

webvictim reviewed (approved):
I'm loath to alter the scope of this PR further, but I recently noticed that we support setting `auth_service.authentication.connector_name` to pass the name of the default connector to use inside Teleport (i.e. `Okta`, `azure-ad` etc)

I think it would make a nice addition as the `authentication.connectorName` value here to round off the functionality.

ptgott commented on docs/pages/setup/helm-reference/teleport-cluster.mdx:
```diff
@@ -63,13 +63,26 @@ This reference details available values for the `teleport-cluster` chart.
-## `authenticationType`
+## `authentication`
+
+### `authentication.type`
```
I think all field-specific headings are H2s.
```suggestion
## `authentication.type`
```

hugoShaka replied:
I checked and we use h3 in the reference for nested values, I'll correct all other titles

ptgott commented on docs/pages/setup/helm-reference/teleport-cluster.mdx:
```diff
@@ -63,13 +63,26 @@ This reference details available values for the `teleport-cluster` chart.
+| `bool` | `true` | No | `auth_service.authentication.local_auth` | ❌ |
+
+`authentication.localAuth` controls whether local authentication is enabled.
+When disabled, users can only login through non-local auth like `saml`, `oidc` or `github`.
```
```suggestion
When disabled, users can only log in through authentication connectors like `saml`, `oidc` or `github`.
```
"login" is a noun. Also making the language consistent with the rest of our docs.

hugoShaka commented:
I'll switch this PR to draft, when working on what @webvictim suggested I found it was not backward compatible for users managing authentication through `cluster_auth_preference` resources.

I think breaking backward compatibility (only for a few users) here will be a good thing as it will allow us to align the helm values with the configuration fields.

I'll submit something else for review in a couple of days, sadly this won't be backportable to v10.

hugoShaka committed:
fixup! helm: allow to disable local auth + nest auth values (#12974)

hugoShaka commented:
The PR is ready for another review. It changed a lot since last review so please check it again.

This PR is backward compatible. I was afraid to break backward compatibility for users relying on the `cluster_auth_preference` resource but after more rigorous testing I realized the Helm chart never supported those users in the first place (except in custom mode). 

Noticeable changes since last review:
- `secondFactor` and `webauthn` are now under `authentication` (if previous values are used they will take precedence)
- `connectorName` support: allows full passwordless setups
- `lockingMode` support: it took me 5 additional minutes and I won't have to create a PR for this feature in 6 months

The only `authentication` field I did not brought back in the chart is `passwordless`. We want to keep Teleport's default behaviour and there is no easy way of having a troolean in yaml/golang/helm. We can wait and see if someone ever needs to manually set the feature flag.

ptgott commented on docs/pages/setup/helm-reference/teleport-cluster.mdx:
```diff
@@ -63,51 +63,82 @@ This reference details available values for the `teleport-cluster` chart.
+| `string` | `""` | No | `auth_service.authentication.connector_name` | ❌ |
+
+`authentication.connectorName` sets the default authentication connector.
+See [the SSO documentation](../../enterprise/sso.mdx). Other possible values are `local` and
```
What are readers supposed to see in the SSO documentation? I would indicate that in this description. 

Since there are only a few SSO connectors, I think it would also make sense to list the possible values here, rather than referring the reader to another docs page.

hugoShaka replied:
This field is either the authentication connector name or `local`/`passwordless`. IIUC SSO connector name != supported SSO providers. Users can give any name they want to their SSO connector and create multiple connectors for the same provider. Hence I cannot list all possible values

ptgott commented on docs/pages/setup/helm-reference/teleport-cluster.mdx:
```diff
@@ -63,51 +63,82 @@ This reference details available values for the `teleport-cluster` chart.
+
+`authentication.connectorName` sets the default authentication connector.
+See [the SSO documentation](../../enterprise/sso.mdx). Other possible values are `local` and
+[`passwordless`](../../access-controls/guides/passwordless.mdx#optional-enable-passwordless-by-default).
```
Should the other values have links as well?

hugoShaka replied:
Done.

ptgott commented on docs/pages/setup/helm-reference/teleport-cluster.mdx:
```diff
@@ -63,51 +63,82 @@ This reference details available values for the `teleport-cluster` chart.
 
-## `authenticationSecondFactor`
+`authentication.lockingMode` controls the locking mode cluster-wide.
+See [the locking modes documentation](../../access-controls/guides/locking.mdx#next-steps-locking-modes)
```
Since there are only two possible values, I think we can list/describe these here, rather than referring the user to another docs page.

hugoShaka replied:
Done

hugoShaka committed:
Doc review

hugoShaka committed:
fixup! Doc review

r0mant commented on examples/chart/teleport-cluster/values.yaml:
```diff
@@ -18,13 +18,52 @@ kubeClusterName: ""
+  localAuth: true
+
+  # Controls the locking mode: in case of network split should Teleport guarantee availability or integrity ?
+  # See https://goteleport.com/docs/access-controls/guides/locking/#next-steps-locking-modes.
```
Nit: I'd include possible values in the comment as well.

hugoShaka replied:
Done

r0mant commented on examples/chart/teleport-cluster/values.yaml:
```diff
@@ -18,13 +18,52 @@ kubeClusterName: ""
+
+  # Enable/disable local authentication by setting `authentication.local_auth` in `teleport.yaml`.
+  # Disabling local auth is required for FedRAMP / FIPS; see https://gravitational.com/teleport/docs/enterprise/ssh-kubernetes-fedramp/
+  localAuth: true
```
So if this setting is `true` by default, what will happens to clusters that have `allow_local_auth` disabled in their ClusterAuthPreference resource after upgrade? Wasn't it the original issue that config setting will take precedence?

hugoShaka replied:
Today you cannot use `cluster_auth_preference` with clusters deployed with the chart. We are always specifying `authentication.type` in `teleport.yaml`. All `cap` resources are flushed on every pod restart.

hugoShaka committed:
List possible lockingMode params

hugoShaka committed:
Merge branch 'master' into hugo/12974-helm-disable-local-auth
